File: packages/brmem/src/brmem/check.py
"""Probe whether a Branch Memory Entry exists."""

from __future__ import annotations

from pathlib import Path
from typing import Annotated

import click

from asdl_core.clinkr.ensure import Ensure
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from brmem.context import load_brmem_context
from brmem.key_validation import check_key
from brmem.ref_layout import (
    EntryRef,
    check_branch_name,
    check_namespace,
    namespace_value_label,
    normalize_namespace_option,
    ref_name_for_entry,
)
from brmem.validation import first_failure


class CheckRequest(ClinkrModel):
    key: Annotated[
        str,
        click.Argument(["key"], type=click.STRING),
    ]
    namespace: Annotated[
        str | None,
        click.Option(
            ["--namespace"],
            type=click.STRING,
            default=None,
            help=("Namespace (e.g. 'notes'). Omit for ad-hoc base Entries."),
        ),
    ] = None
    branch: str | None = None
    at: str | None = None


class CheckResult(ClinkrModel):
    namespace: str
    key: str
    branch: str
    ref_name: str
    target: str
    at: str | None
    head_sha: str | None
    head_date: str | None
    blob_sha: str | None
    size_bytes: int | None


def render_check(result: CheckResult) -> None:
    namespace_label = namespace_value_label(result.namespace)
    lines = [
        f"Namespace: {namespace_label}",
        f"Entry Key: {result.key}",
        f"Branch: {result.branch}",
        f"Entry Locator: {result.ref_name}",
        f"Target: {result.target}",
        f"Head: {result.head_sha} ({result.head_date})",
        f"Blob: {result.blob_sha}",
        f"Size: {result.size_bytes}",
    ]
    click.echo("\n".join(lines))


@clinkr_operation(
    name="check",
    help="Check whether a Branch Memory Entry exists.",
    human_renderer=render_check,
)
def run_check(
    ctx: click.Context,
    request: CheckRequest,
) -> ClinkrExit[CheckResult]:
    brmem_context = load_brmem_context(ctx)
    branch = (
        request.branch
        if request.branch is not None
        else Ensure.ideal_state(brmem_context.git_gateway.get_current_branch(Path.cwd()))
    )

    namespace = normalize_namespace_option(request.namespace)

    failure = first_failure(
        ("invalid_namespace", check_namespace(namespace)),
        ("invalid_key", check_key(request.key)),
        ("invalid_branch_name", check_branch_name(branch)),
    )
    error_type, message = failure or ("", "")
    Ensure.true(
        failure is None,
        error_type=error_type,
        message=message,
    )

    entry_ref = EntryRef(
        namespace=namespace,
        key=request.key,
        branch=branch,
        ref_name=ref_name_for_entry(namespace, request.key, branch),
    )

    target = request.at if request.at is not None else entry_ref.ref_name

    diagnostic = brmem_context.brmem_gateway.check(
        entry_ref.namespace,
        entry_ref.key,
        entry_ref.branch,
        at=request.at,
    )

    if diagnostic is None:
        absent = CheckResult(
            namespace=entry_ref.namespace,
            key=entry_ref.key,
            branch=entry_ref.branch,
            ref_name=entry_ref.ref_name,
            target=target,
            at=request.at,
            head_sha=None,
            head_date=None,
            blob_sha=None,
            size_bytes=None,
        )
        namespace_label = namespace_value_label(entry_ref.namespace)
        raise ClinkrExit.negative(
            absent,
            message=(
                f"not found: Entry Key={entry_ref.key} Namespace={namespace_label} "
                f"Branch={entry_ref.branch} at {target}"
            ),
        )

    return ClinkrExit.ok(
        CheckResult(
            namespace=entry_ref.namespace,
            key=entry_ref.key,
            branch=entry_ref.branch,
            ref_name=entry_ref.ref_name,
            target=target,
            at=request.at,
            head_sha=diagnostic.head_sha,
            head_date=diagnostic.head_date,
            blob_sha=diagnostic.blob_sha,
            size_bytes=diagnostic.size_bytes,
        )
    )

File: ts/packages/brmem/src/operations/check.ts
import { negative, ok } from "@asdl/clinkr";
import { z } from "zod";

import type { BrmemCliContext } from "../context.ts";
import { mustEntryLocator, namespaceValueLabel } from "../ref-layout.ts";
import { gatewayFailure, resolveEntryRequest } from "./shared.ts";

export const checkRequestSchema = z.object({
	key: z.string().describe("Entry Key to check."),
	namespace: z.string().optional().describe("Namespace. Omit for ad-hoc base Entries."),
	branch: z.string().optional().describe("Branch. Defaults to current branch."),
	at: z.string().optional().describe("Snapshot ref or commit to inspect."),
});

export const checkResultSchema = z.object({
	namespace: z.string(),
	key: z.string(),
	branch: z.string(),
	ref_name: z.string(),
	target: z.string(),
	at: z.string().nullable(),
	head_sha: z.string().nullable(),
	head_date: z.string().nullable(),
	blob_sha: z.string().nullable(),
	size_bytes: z.number().int().nullable(),
});

export type CheckRequest = z.infer<typeof checkRequestSchema>;
export type CheckResult = z.infer<typeof checkResultSchema>;

export async function runCheck(ctx: BrmemCliContext, request: CheckRequest) {
	const resolved = await resolveEntryRequest(ctx, request);
	if (resolved.type !== "resolved") return resolved;
	const { namespace, key, branch } = resolved.value;
	const locator = mustEntryLocator(namespace, key, branch);
	const target = request.at ?? locator;
	const result = await ctx.gateway.checkEntry({ namespace, key, branch, at: request.at });
	if (result.type === "error") return gatewayFailure<CheckResult>(result.error);
	if (result.type === "missing") {
		const absent = emptyResult({ namespace, key, branch, refName: locator, target, at: request.at });
		return negative(
			`not found: Entry Key=${key} Namespace=${namespaceValueLabel(namespace)} Branch=${branch} at ${target}`,
			absent,
		);
	}
	return ok({
		namespace,
		key,
		branch,
		ref_name: locator,
		target,
		at: request.at ?? null,
		head_sha: result.value.headSha,
		head_date: result.value.headDate,
		blob_sha: result.value.blobSha,
		size_bytes: result.value.sizeBytes,
	});
}

export function renderCheck(result: CheckResult): string {
	const lines = [
		`Namespace: ${namespaceValueLabel(result.namespace)}`,
		`Entry Key: ${result.key}`,
		`Branch: ${result.branch}`,
		`Entry Locator: ${result.ref_name}`,
		`Target: ${result.target}`,
		`Head: ${result.head_sha} (${result.head_date})`,
		`Blob: ${result.blob_sha}`,
		`Size: ${result.size_bytes}`,
	];
	return lines.join("\n");
}

function emptyResult(options: {
	namespace: string;
	key: string;
	branch: string;
	refName: string;
	target: string;
	at?: string | undefined;
}): CheckResult {
	return {
		namespace: options.namespace,
		key: options.key,
		branch: options.branch,
		ref_name: options.refName,
		target: options.target,
		at: options.at ?? null,
		head_sha: null,
		head_date: null,
		blob_sha: null,
		size_bytes: null,
	};
}

